File: src/commands/investigate.ts
import { Command } from 'commander'
import chalk from 'chalk'
import { getGlobalOpts, output, error, success, info } from '../lib/output.js'
import { listTrackedRepos, listCodeCommitRepos, addTrackedRepo } from '../lib/aws.js'

interface TemporalStartResult {
  workflowId: string
  runId?: string
}

async function startTemporalWorkflow(
  temporalUrl: string,
  namespace: string,
  taskQueue: string,
  workflowId: string,
  workflowType: string,
  args: unknown[]
): Promise<TemporalStartResult> {
  const response = await fetch(`${temporalUrl}/api/v1/namespaces/${namespace}/workflows/${encodeURIComponent(workflowId)}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      workflowType: { name: workflowType },
      taskQueue: { name: taskQueue },
      input: { payloads: args.map(a => ({ data: Buffer.from(JSON.stringify(a)).toString('base64'), metadata: { encoding: 'anNvbi9wbGFpbg==' } })) }
    })
  })
  if (!response.ok) {
    const text = await response.text()
    throw new Error(`Temporal API error (${response.status}): ${text}`)
  }
  const data = await response.json() as { runId?: string }
  return { workflowId, runId: data.runId }
}

export const investigateCommand = new Command('investigate')
  .description('Trigger architecture investigation')
  .argument('[repo]', 'Repository name (omit for all)')
  .option('--all', 'Investigate all enabled repos')
  .option('--discover', 'Auto-discover repos before investigating')
  .option('--model <model>', 'Model ID', 'us.anthropic.claude-sonnet-4-6')
  .option('--chunk-size <n>', 'Chunk size', '10')
  .option('--parallel <n>', 'Parallel limit (daily only)', '3')
  .option('--temporal-url <url>', 'Temporal HTTP API URL', 'http://temporal-alb-internal:8233')
  .option('--namespace <ns>', 'Temporal namespace', 'default')
  .option('--task-queue <queue>', 'Temporal task queue', 'investigate-task-queue')
  .action(async (repo, opts, cmd) => {
    const global = getGlobalOpts(cmd)
    try {
      // Auto-discover if requested
      if (opts.discover) {
        info('Auto-discovering repos from CodeCommit...')
        const discovered = await listCodeCommitRepos(global)
        const existing = await listTrackedRepos(global)
        const existingNames = new Set(existing.map(r => r.name))
        let added = 0
        for (const r of discovered) {
          if (!existingNames.has(r.name)) {
            await addTrackedRepo(global, r)
            added++
          }
        }
        if (added > 0) info(`Discovered ${discovered.length} repos, added ${added} new`)
      }

      if (repo) {
        // Single repo investigation
        const workflowId = `investigate-single-${repo}-${Date.now()}`
        const result = await startTemporalWorkflow(
          opts.temporalUrl, opts.namespace, opts.taskQueue,
          workflowId, 'InvestigateSingleRepoWorkflow',
          [{ repoName: repo, model: opts.model, chunkSize: parseInt(opts.chunkSize) }]
        )
        output(cmd, { success: true, ...result, repo }, () => {
          success(`Investigation started for ${chalk.bold(repo)}`)
          console.log(`  Workflow: ${result.workflowId}`)
        })
      } else if (opts.all) {
        // Daily/all repos investigation
        const repos = await listTrackedRepos(global)
        const enabled = repos.filter(r => r.enabled)

        if (enabled.length === 0) {
          error(cmd, 'No enabled repos. Run `reposwarm discover` or `reposwarm investigate --all --discover`')
          return
        }

        const repoNames = enabled.map(r => r.name)
        const workflowId = `investigate-daily-${Date.now()}`
        const result = await startTemporalWorkflow(
          opts.temporalUrl, opts.namespace, opts.taskQueue,
          workflowId, 'InvestigateReposWorkflow',
          [{ repos: repoNames, model: opts.model, chunkSize: parseInt(opts.chunkSize), parallelLimit: parseInt(opts.parallel) }]
        )
        output(cmd, { success: true, ...result, repoCount: repoNames.length, repos: repoNames }, () => {
          success(`Daily investigation started for ${chalk.bold(String(repoNames.length))} repos`)
          console.log(`  Workflow: ${result.workflowId}`)
        })
      } else {
        error(cmd, 'Specify a repo name or use --all. Example: reposwarm investigate my-repo')
      }
    } catch (err) {
      error(cmd, 'Investigation failed', err)
    }
  })

File: internal/commands/investigate.go
package commands

import (
	"fmt"

	"github.com/loki-bedlam/reposwarm-cli/internal/api"
	"github.com/loki-bedlam/reposwarm-cli/internal/config"
	"github.com/loki-bedlam/reposwarm-cli/internal/output"
	"github.com/spf13/cobra"
)

func newInvestigateCmd() *cobra.Command {
	var model string
	var chunkSize, parallel int
	var all bool

	cmd := &cobra.Command{
		Use:   "investigate [repo]",
		Short: "Trigger architecture investigation",
		Long: `Trigger an AI-powered architecture investigation for one or all repos.

Examples:
  reposwarm investigate is-odd              # Single repo
  reposwarm investigate --all               # All enabled repos
  reposwarm investigate is-odd --model us.anthropic.claude-opus-4-6`,
		RunE: func(cmd *cobra.Command, args []string) error {
			client, err := getClient()
			if err != nil {
				return err
			}

			cfg, _ := config.Load()
			if model == "" {
				model = cfg.DefaultModel
			}
			if chunkSize == 0 {
				chunkSize = cfg.ChunkSize
			}

			if len(args) > 0 {
				// Single repo
				req := api.InvestigateRequest{
					RepoName:  args[0],
					Model:     model,
					ChunkSize: chunkSize,
				}
				var result any
				if err := client.Post(ctx(), "/investigate/single", req, &result); err != nil {
					return err
				}
				if flagJSON {
					return output.JSON(result)
				}
				output.Successf("Investigation started for %s", output.Bold(args[0]))
				return nil
			}

			if all {
				req := api.InvestigateDailyRequest{
					Model:         model,
					ChunkSize:     chunkSize,
					ParallelLimit: parallel,
				}
				var result any
				if err := client.Post(ctx(), "/investigate/daily", req, &result); err != nil {
					return err
				}
				if flagJSON {
					return output.JSON(result)
				}
				output.Successf("Daily investigation started for all enabled repos")
				return nil
			}

			return fmt.Errorf("specify a repo name or use --all\n\nExamples:\n  reposwarm investigate my-repo\n  reposwarm investigate --all")
		},
	}

	cmd.Flags().BoolVar(&all, "all", false, "Investigate all enabled repos")
	cmd.Flags().StringVar(&model, "model", "", "Model ID (default from config)")
	cmd.Flags().IntVar(&chunkSize, "chunk-size", 0, "Files per chunk (default from config)")
	cmd.Flags().IntVar(&parallel, "parallel", 3, "Parallel limit (daily only)")
	return cmd
}
